### PR TITLE
Ethan's changes to handle columns with degeneracies

### PR DESCRIPTION
**Pull request by ekikinzon**
Pulled from Ethan's fork

### CODE REVIEW

File: src/par/FixColumnPartitions.c [hunk 1]
```diff
@@ -19,138 +19,162 @@
...
+  List_ptr fregs;
+  int* touched;
+  FILE* fid;
+  
+  if (MESH_Num_Regions(mesh) == 0) return 1; /* Not implemented for 2D (perhaps not needed either) */
+
+  /* ensure every region is touched once */
+  touched = calloc(MESH_Num_Regions(mesh),sizeof(int));
+  for (idx=0; idx!=MESH_Num_Regions(mesh); ++idx) touched[idx] = 0;
+  printf("Read %-d regions", MESH_Num_Regions(mesh));
```
Comment (raovgarimella) [inline src/par/FixColumnPartitions.c]: Can we put this within #ifdef DEBUG

File: src/par/FixColumnPartitions.c [hunk 2]
```diff
@@ -19,138 +19,162 @@
...
 
+  /* loop over all regions, finding regions whose "up" face is a boundary */
+  fid = fopen("col_counts.txt", "w");
```
Comment (raovgarimella) [inline src/par/FixColumnPartitions.c]: Likewise, can we write out a file only if #ifdef DEBUG is enabled

File: src/par/FixColumnPartitions.c [hunk 3]
```diff
@@ -19,138 +19,162 @@
...
+
+  num_cols = 0;
+  num_touches = 0;
+  idx = 0;
+  while ((mr = MESH_Next_Region(mesh,&idx))) {
+    FixColumnPartitions_UpDown(mesh, mr, &topf, &botf);
+
```
Comment (raovgarimella) [inline src/par/FixColumnPartitions.c]: Check if you really got a topf and botf - otherwise code will crash inelegantly

Comment (raovgarimella) [reply src/par/FixColumnPartitions.c]: Move on to next region if you didn't get both

File: src/par/FixColumnPartitions.c [hunk 4]
```diff
@@ -19,138 +19,162 @@
...
+    while (!done) {
+      if (touched[curid-1]) 
+        MSTK_Report("FixColumnPartitions","Mesh is not columnar, iteration process touches the same cell twice.",MSTK_FATAL);
+      touched[curid-1] = 1;
+      num_cells_in_col++;
+      num_touches++;
+      
+      /* find the bottom face of that region */
+      FixColumnPartitions_UpDown(mesh, curreg, &topf2, &botf);
+      if (topf2 != topf)
```
Comment (raovgarimella) [inline src/par/FixColumnPartitions.c]: Maybe use "```expected_topf```" and "```topf```" and a call of ```FixColumnPartitions(mesh, curreg, &topf, &botf)```. Right now, the first reaction when we see ```if (topf2 != topf)``` is Whaaaa???

File: src/par/FixColumnPartitions.c [hunk 5]
```diff
@@ -162,20 +186,38 @@ int FixColumnPartitions(Mesh_ptr mesh, int *part, MSTK_Comm comm) {
...
   if (nmove) {
     char msg[256];
-    sprintf(msg,"Redistributed %-d elements to maintain column partitioning",nmove);
+    sprintf(msg,"Redistributed %-d elements in %-d columns to maintain column partitioning",nmove, num_cols);
     MSTK_Report("FixColumnPartitions",msg,MSTK_MESG);
   }
-    
+
+  fclose(fid);
+  num_not_touched = 0;
```
Comment (raovgarimella) [inline src/par/FixColumnPartitions.c]: What if part of the mesh is columnar and the rest is not? Maybe the message should say that algorithm could not put all elements in a column - perhaps part of the mesh is not columnar?

File: src/par/FixColumnPartitions.c [hunk 6]
```diff
@@ -19,138 +19,162 @@
...
 
-int FixColumnPartitions(Mesh_ptr mesh, int *part, MSTK_Comm comm) { 
-  int i, j, k, idx, nrf, nfr, nfv, nrv, curid, nxtid, homepid, nxtpid;
-  int interior, done, nmove=0, found;
-  MRegion_ptr mr, curreg, nxtreg;
-  MFace_ptr rf, topf, botf;
-  MVertex_ptr fv;
-  List_ptr rfaces, fregs, fverts;
-  double fxyz[MAXPV2][3], rxyz[MAXPV3][3], rfvec[3], rcen[3], fcen[3], maxdp, norm;
+int FixColumnPartitions_IsSide(Mesh_ptr mesh, MRegion_ptr mr, MFace_ptr rf) {
```
Comment (raovgarimella) [inline src/par/FixColumnPartitions.c]: Rename to ```FixColumnPartitions_IsSideFace``` - just a suggestion

File: src/par/FixColumnPartitions.c [hunk 7]
```diff
@@ -19,138 +19,162 @@
...
-    
-    /* Hopefully, found the top face of a column in the mesh */
+  List_Delete(rfaces);
+  return ret;
+}
 
-    curreg = List_Entry(fregs,0);
-    curid = MR_ID(curreg);
-    homepid = part[curid-1];
+
```
Comment (raovgarimella) [inline src/par/FixColumnPartitions.c:119]: There is some repetition of work in this routine (as there was in the previous version). If this is an issue, one can go through all faces and mark them as being "side" faces or not (using a temporary array) and then use that to zoom into the up and down faces in the FixColumnPartions_UpDown

File: src/par/FixColumnPartitions.c [hunk 8]
```diff
@@ -19,138 +19,162 @@
...
 
-    rcen[0] = rcen[1] = rcen[2] = 0.0;
-    MR_Coords(mr,&nrv,rxyz);
-    for (i = 0; i < nrv; i++) {
-      for (k = 0; k < 3; k++)
-        rcen[k] += rxyz[i][k];
+   Returns 1 if the relationship is proper, or 0 if the region has
+   zero volume and is therefore uncertain.
+*/
+int FixColumnPartitions_UpDown(Mesh_ptr mesh, MRegion_ptr mr, MFace_ptr* up, MFace_ptr* dn) {
```
Comment (raovgarimella) [inline src/par/FixColumnPartitions.c]: FixColumnParitions_UpDownFaces? - just a suggestion

**Review by raovgarimella — CHANGES_REQUESTED** [review]
Some minor tweaks to make. Also, is there a test that was failing with the old code and passes now with the new? It would be good to put that in

File: src/par/FixColumnPartitions.c (hunk 3 above)
Comment (ekikinzon) [reply src/par/FixColumnPartitions.c]: Actually, the FixColumnPartitions_UpDown is going to throw an error if either of top or botf is NULL:

`  /* find two faces that are not sides */
  found = 0; i = 0;
  while (found < 2) {
    rf = List_Entry(rfaces,i);
    if (!FixColumnPartitions_IsSide(mesh, mr, rf)) {
      if (found < 1) *up = rf;
      else *dn = rf;
      found++;
    }
    i++;
    if (i >= nrf) 
      MSTK_Report("FixColumnPartitions","Mesh is not columnar, can't find both up and down faces.",MSTK_FATAL);
  }`

This also implies that we _have_ to get both, i.e. the function will throw an error if any part of domain can be represented as columns.

**Review by raovgarimella — APPROVED** [review]

**Pull request merged**